craft_parts/utils/layout_utils.py

```diff
@@ -0,0 +1,73 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2025 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Layout utilities."""
+
+from typing import Any
+
+from craft_parts import errors, features
+from craft_parts.layouts import LayoutItem
+
+
+def validate_layout(data: list[dict[str, Any]]) -> None:
+    """Validate a layout.
+
+    :param data: The repository data to validate.
+    """
+    if not isinstance(data, list):  # pyright: ignore[reportUnnecessaryIsInstance]
+        raise TypeError("value must be a dictionary")
+    layout = [LayoutItem.unmarshal(item) for item in data]
+
+    if layout[0].mount != "/":
+        raise ValueError("A filesystem first entry must map the '/' mount.")
+
+
+def validate_layouts(layouts: dict[str, Any] | None) -> None:
+    """Validate the layouts section.
+
+    If layouts are defined then both partition and overlay features must
+    be enabled.
+    A layout dict must only have a single "default" entry.
+    The first entry in default must map the '/' mount.
+    """
+    if not layouts:
+        return
+
+    if (
+        not features.Features().enable_partitions
+        or not features.Features().enable_overlay
+    ):
+        raise errors.FeatureError(
+            "Filesystems are defined but partition feature or overlay feature are not enabled."
```

Comment:
Where is the line drawn for when to use filesystems instead of layouts? Is it similar to the problem we have with craft-parts raising errors about partitions, when a Snapcraft user has an error with components?

If so, one solution would be to create a new class of errors, LayoutError. That way imagecraft could catch LayoutErrors and raise an error that talks specifically about filesystems.

Reply:
The "layout" term is for us, to avoid confusion when working with it in craft-parts/craft-applications and *craft applications. This is indeed similar to the problem with partitions because I think users should only be aware of "filesystems". However I expect that users will always interact with the "filesystems" concept, whatever the application, and not with an application-specific concept (components for Snapcraft, disk partitions for Imagecraft, etc.) so in a way this is simpler than the "components" problem. So I am not sure we need to translate the errors raised in craft-parts in the *craft applications. 

I am not satisfied with using `FeatureError` though, so I will try introducing a `LayoutError` as you suggested.